### PR TITLE
fix(ci): prevent smoke test timeout cancellation

### PR DESCRIPTION
**Pull request by SebTardif**
## Problem

The Smoke Tests workflow has a 15-minute job timeout that is too tight. Build steps take ~80s, and the shell retry loop (`max_attempts=3`) reruns **all 8 scenarios** from scratch on any failure. With each parallel run taking ~6 minutes, three full-suite retries total ~18 minutes of test time, exceeding the 15-minute budget.

This causes the job to be cancelled with "The operation was canceled" — not because of a test failure, but because the timeout kills it mid-run. This was observed on PR #10525 and other recent PRs.

**Timeline from PR #10525 (commit `13facd7c`):**
- Build steps: ~80s
- Smoke test step: ran for 13 min 51 sec → **cancelled by 15-min job timeout**

### 1. Increase timeout (15 → 30 minutes)
Provides adequate headroom for build + retries, even in worst-case scenarios.

### 2. Replace shell retry with runner's built-in `--trials 3`
The old approach ran `--trials 1` inside a `max_attempts=3` shell loop that retried **all scenarios** on any failure. The new approach uses `--trials 3 --parallel`, which gives each scenario 3 independent chances (same flaky tolerance as before).

| | Old (shell retry) | New (--trials 3) |
|---|---|---|
| Retry scope | Entire suite (all 8 scenarios) | Per-scenario |
| Passing scenarios re-run? | Yes, on every retry | No (early exit) |
| Worst case (all fail) | 3 × 8 scenarios = 24 runs | 8 × 3 trials = 24 runs |
| Best case (all pass) | 8 runs | 8 runs |
| Typical (1 flaky) | 8 + 8 + 8 = 24 runs | 7 + 3 = 10 runs |

### 3. Early-exit optimization in runner
When a trial passes, skip remaining trials for that scenario. This means passing scenarios run exactly once, and only failures use all 3 trials.

## Before vs After

**Before:** `~80s build + up to 18 min tests = ~19 min` → exceeds 15 min timeout → CANCELLED  
**After:** `~80s build + ~6-8 min tests = ~8 min` → well within 30 min timeout

---

> [!NOTE]
> **Low Risk**
> Changes are limited to eval/smoke-test execution and metrics reporting, with no production runtime or auth impact.
> 
> **Overview**
> Adds **per-scenario early exit** in the smoke test runner: when `trials > 1`, it stops after the first passing trial instead of running the full retry budget.
> 
> Introduces **`MetricsCalculator.applyEarlyExitPassAtK`** so reporting still treats pass@k as “found a solution within k attempts” when fewer than the requested trials ran. **pass^k** and **flakiness** stay based only on trials that actually executed—avoiding padded `true` values that would falsely show perfect reliability.
> 
> Unit tests cover early-exit edge cases (no-op when all trials run or last trial fails, pass@3 only when success is within three attempts, anti-regression vs true-padding).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit d6cd03e91192dc34b291da70f077d45c7e4ca257. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: evals/smoke-tests/run-smoke-tests.ts [hunk 1]
```diff
@@ -550,6 +550,9 @@ async function main() {
 			trialWorkdirs.push(trialWorkdir)
 			const result = await runTrial(scenario, modelId, trialWorkdir)
 			trialResults.push(result)
+			if (result.passed && trials > 1) {
+				break
+			}
```
[the remark this answers is not in the document]
Comment (SebTardif) [reply evals/smoke-tests/run-smoke-tests.ts:555]: Fixed in 5afe9f1 — when early-exiting on a pass, `trialBools` is now padded with `true` entries to match the expected `trials` count before passing to `calculateTaskMetrics`. This ensures `passAt3` computes correctly for scenarios that pass on their first trial.

**Comment by SebTardif** [comment]
Rebased onto current monorepo main.

cline-evals-regression.yml no longer exists, so the workflow timeout/retry change from the original PR is dropped. Kept the runner-side improvement in evals/smoke-tests/run-smoke-tests.ts: early-exit on first pass when --trials > 1, with trial-bool padding so pass@k stays consistent.

File: evals/smoke-tests/run-smoke-tests.ts [hunk 2]
```diff
@@ -565,6 +568,9 @@ async function main() {
 		})
 
 		const trialBools = trialResults.map((t) => t.passed)
+		if (trialResults.length < trials && trialResults[trialResults.length - 1]?.passed) {
+			trialBools.push(...Array(trials - trialResults.length).fill(true))
+		}
```
[the remark this answers is not in the document]
Comment (SebTardif) [reply evals/smoke-tests/run-smoke-tests.ts]: @cursor

> Padding `trialBools` with `true` for unrun trials after an early exit, while intended for `passAtK` metrics, incorrectly inflates `passCaret3` to `1.0`, distorts `passAt1`, and sets `flakinessScore` to `0`.

Agreed. Fixed in 3c2710d8f: metrics are computed from trials that actually ran, then `applyEarlyExitPassAtK` sets pass@k only (solution found within k). pass^k and flakiness stay based on observed runs. Unit tests cover the anti-padding regression.

File: evals/analysis/src/metrics.ts [hunk 3]
```diff
@@ -183,4 +183,37 @@ export class MetricsCalculator {
...
+		passAt3: number
+		passCaret3: number
+		flakinessScore: number
+	} {
+		if (trialsRan < trialsRequested && lastTrialPassed) {
+			metrics.passAt1 = 1
+			if (trialsRequested >= 3) {
+				metrics.passAt3 = 1
+			}
+		}
```
[the remark this answers is not in the document]
Comment (SebTardif) [reply evals/analysis/src/metrics.ts:220]: @cursor

> `applyEarlyExitPassAtK` sets `passAt1` and `passAt3` to 1 on any early-exit pass, without checking that the first success occurred within 1 or 3 attempts.

Fixed in 67ed50bfb: overrides are gated on `trialsRan` (1-based first-pass index). `passAt1` only when `trialsRan <= 1`, `passAt3` only when `trialsRan <= 3`. Unit tests cover first success after trial 3 and on trial 2 of 5.

File: evals/analysis/src/__tests__/metrics.test.ts [hunk 4]
```diff
@@ -191,6 +191,71 @@ describe("MetricsCalculator", () => {
...
+			// Fail fail fail pass, then early-exit with trialsRequested > 4
+			const observed = [false, false, false, true]
+			const base = calc.calculateTaskMetrics(observed)
+			const metrics = calc.applyEarlyExitPassAtK(base, 4, 5, true)
+
+			expect(metrics.passAt1).toBe(base.passAt1)
+			expect(metrics.passAt3).toBe(base.passAt3)
+			expect(metrics.passAt3).not.toBe(1.0)
+			expect(metrics.passCaret3).toBe(base.passCaret3)
+		})
```
[the remark this answers is not in the document]
Comment (SebTardif) [reply evals/analysis/src/__tests__/metrics.test.ts:248]: @cursor

> `applyEarlyExitPassAtK` mutates and returns its `metrics` argument, but this test passes `base` without copying.

Fixed in the follow-up commit: the test now spreads `{ ...base }` before calling, matching sibling tests, so assertions against the original `base` can detect accidental mutations.
